addons/html_editor/static/src/main/selection_placeholder_plugin.js

```diff
@@ -0,0 +1,251 @@
+import { Plugin } from "@html_editor/plugin";
+import { closestBlock, isBlock } from "@html_editor/utils/blocks";
+import { fillEmpty } from "@html_editor/utils/dom";
+import {
+    allowsParagraphRelatedElements,
+    isEmpty,
+    isNotEditableNode,
+} from "@html_editor/utils/dom_info";
+import {
+    closestElement,
+    getAdjacentNextSiblings,
+    getAdjacentPreviousSiblings,
+} from "@html_editor/utils/dom_traversal";
+import { withSequence } from "@html_editor/utils/resource";
+
+const blinkerClass = "o-horizontal-caret";
+const placeholderAttribute = "data-selection-placeholder";
+const placeholderSelector = `[${placeholderAttribute}]`;
+/**
+ * @param {"before"|"after"} side
+ * @param {Node} node
+ * @returns {Node|undefined}
+ */
+const getNonWhitespaceSibling = (side, node) => {
+    const siblings =
+        side === "before" ? getAdjacentPreviousSiblings(node) : getAdjacentNextSiblings(node);
+    return siblings.find(
+        (sibling) => !(sibling.nodeType === Node.TEXT_NODE && !sibling.textContent.trim())
+    );
+};
+/**
+ * Get an element's top or bottom margin as a number.
+ *
+ * @param {Element} element
+ * @param {"top"|"bottom"} side
+ * @returns {Number}
+ */
+const getMargin = (element, side) =>
+    +element.ownerDocument.defaultView
+        .getComputedStyle(element)
+        [side === "top" ? "marginTop" : "marginBottom"].replace("px", "");
+
+export class SelectionPlaceholderPlugin extends Plugin {
+    static id = "selectionPlaceholder";
+    static dependencies = ["baseContainer", "history", "selection"];
+    resources = {
+        external_history_step_handlers: this.updatePlaceholders.bind(this),
+        normalize_handlers: this.updatePlaceholders.bind(this),
+        step_added_handlers: this.updatePlaceholders.bind(this),
+        selectionchange_handlers: (selectionData) => this.onSelectionChange(selectionData),
+        clean_for_save_handlers: withSequence(0, ({ root }) => {
+            for (const placeholder of root.querySelectorAll(placeholderSelector)) {
+                placeholder.remove();
+            }
+        }),
+        split_element_block_overrides: ({ blockToSplit }) => {
+            if (blockToSplit.hasAttribute(placeholderAttribute)) {
+                this.persistPlaceholder(blockToSplit);
```

Comment:
I never liked this nomenclature. Why "persist"? Aren't they all persisting anyway? I mean they're always in the DOM, waiting for an arrow up/down, right? Why not something more in the direction of `openPlaceholder`? I'm open to suggestions.

Reply:
I called it "persist" because until they're persisted they're temporary. Meaning the non-persisted placeholders (aka the placeholders) get removed on `cleanForSave`, and are candidates for removal on `updatePlaceholders`. Once they're "persisted"... they persist. They've become real paragraphs, visible and persistent. I like the nomenclature, and certainly prefer it to `openPlaceholder` (I don't really know what it would mean to open a placeholder). I'm open to changing though but I don't see anything better and I think the current term describes it quite aptly.